src/ips.rs

```diff
@@ -0,0 +1,533 @@
+            // located nodes.
+            if self.config.use_geolocation {
+                self.update_rating_by_location(node, &state.nodes, &mut peer_ratings);
+            }
+
+            // Load peerlist with current connections (we don't want to change everything)
+            for peer in &agraph[node_idx] {
+                peer_list_entry
+                    .list
+                    .push(IpAddr::from_str(state.nodes[*peer].ip.as_str()).expect(ERR_PARSE_IP));
+
+                // Remember current peer ratings
+                curr_peer_ratings.push(peer_ratings[*peer]);
+            }
+
+            // Get current node's degree for further computations
+            let degree = *degrees.get(&node_ip).expect(ERR_GET_DEGREE);
+
+            // 2 - Calculate desired vertex degree
+            // In the first iteration we will use degree average so all nodes should pursue to
+            // that level. That could be bad if graph's vertexes have very high (or low) degrees
+            // and therefore, delta is very high (or low) too. But until we have some better idea
+            // this one is the best we can do to keep up with the graph.
+            let desired_degree = ((degree_avg + degree as f64) / 2.0).round() as u32;
+
+            // 3 - Calculate how many peers to add or delete from peerlist
+            let mut peers_to_delete_count = if desired_degree < degree {
+                degree.saturating_sub(desired_degree)
+            } else {
+                // Check if config forces to change peerlist even if we have good degree.
+                // This should be always set to at least one to allow for some changes in graph -
+                // searching for better potential peers.
+                self.config.change_at_least
+            };
+
+            // Calculating how many peers should be added. If we have more peers than desired degree
+            // we will add at least config.change_at_least peers.
+            let mut peers_to_add_count = if desired_degree > degree {
+                desired_degree
+                    .saturating_sub(degree)
+                    .saturating_add(peers_to_delete_count)
+            } else {
+                self.config.change_at_least
+            };
+
+            // Limit number of changes to config value
+            if peers_to_add_count > self.config.change_no_more {
+                peers_to_add_count = self.config.change_no_more;
+            }
+
+            // Sort peers by rating (highest first)
+            curr_peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+            // 4 - Choose peers to delete from peerlist (based on ranking)
+            while peers_to_delete_count > 0 {
+                // Remove peers with lowest rating
+                if let Some(peer_to_delete) = curr_peer_ratings.pop() {
+                    peer_list_entry.list.retain(|x| x != &peer_to_delete.0);
```

Comment:
`pop()` removes the peer so this should suffice, no?
```suggestion
                if let Some(peer_to_delete) = curr_peer_ratings.pop() {
```